src/charm.py

```diff
@@ -481,741 +43,243 @@ def __init__(self, *args) -> None:  # type: ignore[no-untyped-def]
-        self.k8s_auth()
-        return kubernetes.client.NetworkingV1Api()
+            The current effective relation object, None if there are no relation.
+        """
+        if self.model.get_relation("nginx-route") is not None:
+            relation = cast(Relation, self.model.get_relation("nginx-route"))
+            if relation.app is not None and relation.data[relation.app]:
+                return relation
+        elif self.model.get_relation("ingress") is not None:
+            relation = cast(Relation, self.model.get_relation("ingress"))
+            if relation.app is not None and relation.data[relation.app]:
+                return relation
+        return None
 
-    def _report_service_ips(self) -> List[str]:
-        """Report on service IP(s) and return a list of them.
+    def _get_definition_from_relation(self, relation: Relation) -> IngressDefinition:
+        """Get the IngressDefinition from the given relation.
 
-        Returns:
-            A list of service IPs.
+        Args:
+            relation: The source relation object.
+
+        Return:
+            The IngressDefinition corresponding to the provided relation.
         """
-        api = self._core_v1_api()
-        services = api.list_namespaced_service(  # type: ignore[attr-defined]
-            namespace=self._namespace
-        )
-        all_k8s_service_names = [rel._k8s_service_name for rel in self._all_config_or_relations]
-        return [
-            x.spec.cluster_ip for x in services.items if x.metadata.name in all_k8s_service_names
-        ]
+        if relation.name == "nginx-route":
+            definition_essence = IngressDefinitionEssence(
+                model=self.model, config=self.config, relation=relation
+            )
+        else:
+            definition_essence = IngressDefinitionEssence(
+                model=self.model,
+                config=self.config,
+                relation=relation,
+                ingress_provider=self._ingress_provider,
+            )
+        ingress_definition = IngressDefinition.from_essence(definition_essence)
+        return ingress_definition
 
-    def _report_ingress_ips(self) -> List[str]:
+    def _report_ingress_ips(self, ingress: kubernetes.client.V1Ingress) -> List[str]:
         """Report on ingress IP(s) and return a list of them.
 
+        Args:
+            ingress: the target ingress.
+
         Returns:
             A list of Ingress IPs.
         """
-        api = self._networking_v1_api()
+        controller = self._get_ingress_controller(namespace=ingress.metadata.namespace)
         # Wait up to `interval * count` seconds for ingress IPs.
-        count, interval = _report_interval_count(), 1
+        count, interval = 100, 1
+        ips = []
         for _ in range(count):
-            ingresses = api.list_namespaced_ingress(  # type: ignore[attr-defined]
-                namespace=self._namespace
-            )
+            ingresses = controller.list_resource()
```

Comment:
let's change this to have a method to get the IP addresses rather than expose the underlying data structure

Reply:
Updated, thanks!